### DISCUSSION TITLE
v2021.11.8-beta1

### DISCUSSION THREAD

**Discussion by ScratMan** [ScratMan/HASmartThermostat#11] [Announcements]
Development version adding optional min_cycle_duration (default value is 0) to filter short on/off cycles for heating systems requiring a minimum on or off time (like boilers).

**Full Changelog**: https://github.com/ScratMan/HASmartThermostat/compare/2021.11.7...2021.11.8-beta1

<hr /><em>This discussion was created from the release <a href='https://github.com/ScratMan/HASmartThermostat/releases/tag/2021.11.8-beta1'>v2021.11.8-beta1</a>.</em>

**Comment by nagyrobi** [comment]
I made this for scheduling: https://github.com/nagyrobi/home-assistant-configuration-examples/tree/main/heating looks like your component was the missing puzzle.

**Reply by nagyrobi** [reply]
It's a little undershot but it works nice with the 10-minutes minimal cycle:
![kép](https://user-images.githubusercontent.com/1550668/142216452-ee2f6e14-5f9a-4159-b94a-bb85c80f58fb.png)
In the house the feeling is a whole lot better! No more warm and cold waves.

I'll try with an autotune in the next few days. 
How does changing the target temperature affect the autotuning?

**Reply by ScratMan** [reply]
I missed your message, sorry. The Autotune now (in beta 4) waits for a few samples to measure the sampling rate of the sensor, then takes the set point of the thermostat and starts regulating the temperature like a standard hysteresis thermostat to analyze the behavior of the system. No need for a step anymore. And you can't adjust the set point till autotune has finished.

As your sampling rate is high (30s), you should keep the 0.5°C noiseband, and on the last beta release (2021.11.8-beta4) a 6 to 10 hours lookback period gave good results.

**Reply by nagyrobi** [reply]
I will try. Thanks.
Should I remove the `sampling_period` setting during autotune?

**Reply by ScratMan** [reply]
No, you can keep it.

**Comment by nagyrobi** [comment]
Maybe worth considering two cycles, separate cycle off and cycle on?

**Reply by ScratMan** [reply]
Possible, but would it be useful ? The generic thermostat only uses 1 setting for both on and off, and it seems relevant to me.
Do you see cases where two settings would be required ?

**Reply by nagyrobi** [reply]
[Thermal Actuators](https://www.roth-werke.de/de/files/MA_FHS_Stellantrieb_1Watt.pdf) have different operation times for opening/closing. They employ a small heating element inside which acts on a dilatating piece of metal. When power applied to such an actuator, it may take up to 8 minutes to completely open (or close, if it's normally open) depending on the operation environment. When power is removed it takes only about 2-3 minutes to completely close (or open respectively) as it cools down faster than it heats up.

![kép](https://user-images.githubusercontent.com/1550668/142181629-06cd4dc0-7b5f-42f5-a6ee-929b6be7d694.png)

So for such devices, for very accurate control it may make sense to have a possibility to specify different ON and OFF cycles.

Note: It's not (only) about the cycle of the actuator operation, but the cycle of the pump! One has to add a delay to the starting of the pump, as the actuators have to open the valves first, and only after that start the pump. So the minimum cycle may have to be extended with the time needed for the actuator to open the valve, to make sure the pump can start pushing some water in the system too...

**Reply by nagyrobi** [reply]
This could be nice with some advanced calculations, to assume based on the elased time the position of the valves and then wait with the pump a dynamic time... but that's a whole other story.

**Reply by nagyrobi** [reply]
To keep backwards compatibility you could have `min_cycle_duration` as it is now, and add a separate _optional_ `min_off_cycle_duration` too.
And then:
if `min_off_cycle_duration` not specified, then `min_off_cycle_duration` = `min_cycle_duration`.

**Reply by ScratMan** [reply]
Understand. I will update with your proposal with a min_off_cycle_override that would be considered only if min_cycle_duration is set.

**Reply by ScratMan** [reply]
Hello, min_off_cycle_duration is implemented in release 2021.11.8-beta2. I also modified the autotuner trying to improve it for better stability and easier debugging.

**Reply by nagyrobi** [reply]
Thank you.

**Comment by nagyrobi** [comment]
Based on a 24-hour test, this is what I see:
![kép](https://user-images.githubusercontent.com/1550668/142769508-16740fac-1208-4a32-8ef3-c9e6723dd31f.png)
The pink line shows the target temperature being adjusted on schedule.

What parameter do you suggest to be adjusted in order for the ramp up to be faster (=do not stop the heater until it reaches the target temp)? The effect is that all schedules need to be shifted in advance with 1-2 hours...

**Reply by nagyrobi** [reply]
My config currently is:
```
  - platform: smart_thermostat
    name: Fűtés
    heater: switch.kazan
    target_sensor: sensor.atlag_belso_homerseklet
    min_temp: 16
    max_temp: 26
    target_temp_step: 0.1
    min_cycle_duration: '00:08:00'
    min_off_cycle_duration: '00:11:00'
    keep_alive:
      seconds: 60
    away_temp: 20.5
    kp: 180
    ki: 0.050
    kd: 30000
    pwm: 00:10:00
```

**Reply by ScratMan** [reply]
Hi, please try with reduced Kd, by 10% or 20%. This should allow the heating to continue a bit more during the rise of the temperature.

**Reply by ScratMan** [reply]
It also looks like your sensor returned many samples per hour, you may limit the number of samples used for the PID using `sampling_period` parameter to limit to half or third the count, it should smooth the on/off curve.

**Reply by nagyrobi** [reply]
The sensor sends the current temperature value every 30 seconds for precise measurement. What's your recommendation for `sampling_period` in this case?

**Reply by ScratMan** [reply]
It will depend on the max slew rate of the temperature when heater is at 100% for a long period. If it takes 1 hour to do a 1°C step, I would say 10 to 15 minutes for sampling_period should be ok, but if it takes 10 minutes, you should rather set it to 2 minutes

**Comment by edouardkleinhans** [comment]
@ScratMan, do you have any recommendation regarding min_cycle_duration for a gaz boiler ?

**Reply by ScratMan** [reply]
Sorry, but no, I can't recommend a value. You should check the start time of your boiler from relay on to hot water output. The minimum cycle should be over this time.

**Comment by nagyrobi** [comment]
> @ScratMan, do you have any recommendation regarding min_cycle_duration for a gaz boiler ?

For a gas boiler 10 minutes is safe enough as `min_cycle_duration` to avoid damage and also get good heat out of it. That means it's gonna turn on and off _at most_ 6 times per hour (3 times on and 3 times off).

**Reply by edouardkleinhans** [reply]
so i need to set min_cycle_duration to 10 minutes. But what about PWM, do i need to set it to 10 minutes too ?

**Reply by nagyrobi** [reply]
I have set mine that to 25 minutes.

**Reply by ScratMan** [reply]
The code will need an update to improve modulation with minimum cycle duration, so that the on time is adjusted when `control_output * PWM / 100 ≥ min_cycle_duration `; and fix on time to `min_cycle_duration` and adjust PWM duration (off time) when `control_output * PWM / 100 < min_cycle_duration`
